configtx/orderer.go

```diff
@@ -185,6 +195,288 @@ func (o *OrdererGroup) Configuration() (Orderer, error) {
+	}
+
+	state := orderer.ConsensusState(ob.ConsensusType_State_name[int32(consensusTypeProto.State)])
+
+	switch consensusTypeProto.Type {
+	case orderer.ConsensusTypeSolo:
+	case orderer.ConsensusTypeKafka:
+		kafkaBrokersValue, ok := o.ordererGroup.Values[orderer.KafkaBrokersKey]
+		if !ok {
+			return errors.New("unable to find kafka brokers for kafka orderer")
+		}
+
+		kafkaBrokersProto := &ob.KafkaBrokers{}
+		err := proto.Unmarshal(kafkaBrokersValue.Value, kafkaBrokersProto)
+		if err != nil {
+			return fmt.Errorf("unmarshaling kafka brokers: %v", err)
+		}
+
+		kafka.Brokers = kafkaBrokersProto.Brokers
+	case orderer.ConsensusTypeEtcdRaft:
+		etcdRaft, err = unmarshalEtcdRaftMetadata(consensusTypeProto.Metadata)
+		if err != nil {
+			return fmt.Errorf("unmarshaling etcd raft metadata: %v", err)
+		}
+	default:
+		return fmt.Errorf("config contains unknown consensus type '%s'", consensusTypeProto.Type)
+	}
+
+	var consensusMetadata []byte
+
+	switch ordererType {
+	case orderer.ConsensusTypeSolo:
+	case orderer.ConsensusTypeKafka:
+		err = setValue(o.ordererGroup, kafkaBrokersValue(kafka.Brokers), AdminsPolicyKey)
+		if err != nil {
+			return err
+		}
+	case orderer.ConsensusTypeEtcdRaft:
+		if consensusMetadata, err = marshalEtcdRaftMetadata(etcdRaft); err != nil {
+			return fmt.Errorf("marshaling etcdraft metadata for orderer type '%s': %v", orderer.ConsensusTypeEtcdRaft, err)
+		}
+	default:
+		return fmt.Errorf("unknown orderer type '%s'", ordererType)
+	}
+
+	consensusState, ok := ob.ConsensusType_State_value[string(state)]
+	if !ok {
+		return fmt.Errorf("unknown consensus state '%s'", state)
+	}
+
+	return setValue(o.ordererGroup, consensusTypeValue(ordererType, consensusMetadata, consensusState), AdminsPolicyKey)
+}
+
+// SetConsensusState sets the consensus state.
+func (o *OrdererGroup) SetConsensusState(consensusState orderer.ConsensusState) error {
+	consensusTypeProto := &ob.ConsensusType{}
+	err := unmarshalConfigValueAtKey(o.ordererGroup, orderer.ConsensusTypeKey, consensusTypeProto)
+	if err != nil {
+		return err
+	}
+
+	return setValue(o.ordererGroup, consensusTypeValue(consensusTypeProto.Type, consensusTypeProto.Metadata, ob.ConsensusType_State_value[string(consensusState)]), AdminsPolicyKey)
+}
+
+// SetTickInterval sets the Etcdraft's tick interval.
+func (e *EtcdRaftOptionsValue) SetTickInterval(interval string) error {
+	consensusTypeProto := &ob.ConsensusType{}
+	err := proto.Unmarshal(e.value.Value, consensusTypeProto)
+	if err != nil {
+		return err
+	}
+
+	etcdRaft, err := unmarshalEtcdRaftMetadata(consensusTypeProto.Metadata)
+	if err != nil {
+		return fmt.Errorf("unmarshaling etcd raft metadata: %v", err)
+	}
+
+	etcdRaft.Options.TickInterval = interval
+
+	consensusMetadata, err := marshalEtcdRaftMetadata(etcdRaft)
+	if err != nil {
+		return fmt.Errorf("marshaling etcdraft metadata for orderer type '%s': %v", orderer.ConsensusTypeEtcdRaft, err)
+	}
+
+	consensusTypeProto.Metadata = consensusMetadata
```

Comment:
This unmarshalling of the ConsensusType to retrieve the underlying orderer.EtcdRaft could probably be moved to a separate function since we repeat it below as well. Maybe a 
```
func (e *EtcdRaftOptions) etcdRaftConfig() (orderer.EtcdRaft, error) {
	consensusTypeProto := &ob.ConsensusType{}
	err := proto.Unmarshal(e.value.Value, consensusTypeProto)
	if err != nil {
		return orderer.EtcdRaft{}, err
	}

	return unmarshalEtcdRaftMetadata(consensusTypeProto.Metadata)
}
```
and then in the Set functions you can call:
```
func (e *EtcdRaftOptionsValue) SetTickInterval(interval string) error {
	etcdRaft, err := e.etcdRaftConfig()
	if err != nil {
		return fmt.Errorf("unmarshaling etcd raft metadata: %v", err)
	}

	etcdRaft.Options.TickInterval = interval
...
}
```

Reply:
Likewise the setting of the option could be added as a helper function as well:
```
function (e *EtcdRaftOptionsValue) setEtcdRaftConfig(etcdRaft orderer.EtcdRaft) error {
	consensusMetadata, err := marshalEtcdRaftMetadata(etcdRaft)
	if err != nil {
		return fmt.Errorf("marshaling etcdraft metadata for orderer type '%s': %v", orderer.ConsensusTypeEtcdRaft, err)
	}

	consensusTypeProto.Metadata = consensusMetadata

        e.value.Value, err = proto.Marshal(consensusTypeProto)
        return err
}
```

making the final Set function call look more like:
```
func (e *EtcdRaftOptionsValue) SetTickInterval(interval string) error {
	etcdRaft, err := e.etcdRaftConfig()
	if err != nil {
		return fmt.Errorf("unmarshaling etcd raft metadata: %v", err)
	}

	etcdRaft.Options.TickInterval = interval
         
        return e.setEtcdRaftConfig(etcdRaft)
}
```
We ultimately probably should have done something similar for the Add/Remove Consenters but that's out of scope for this.